### ISSUE TITLE
Unable to generate samples

### ISSUE THREAD

**Issue by adiosbyebye** [andresguzco/ef-vfm#1]
Hi there, I recently came across this awesome paper and was trying to recreate the results. However there are a few issues. Firstly, running the vanilla code prompts a matrix multiplication dimension mismatch error. The problematic line is line 91 in flow_model.py:
`d_in = self.num_numerical_features + sum(self.num_classes) + len(self.num_classes)`

Changing it to 
`d_in = self.num_numerical_features + sum(self.num_classes)`

works. If you could verify this issue doesnt break the code logic that would be great.

**More critically**, during testing both for datasets "adult" and "default" the code gets stuck at:
`trajectory = odeint(vf, x0, t, method="dopri5", rtol=1e-3, atol=1e-3)`
modifying batch_sample in the config file doesnt fix this.

I tried to recreate the sampling process using a a fixed-step RK4 process via

` 

    def sample(self, num_samples):
        
        dev = self.device
        dt = torch.float32
        d_in = self.num_numerical_features + sum(self.num_classes)  # ODE state dim
        d_out = self.num_numerical_features + len(self.num_classes)  # returned sample dim

        x0 = torch.randn(num_samples, d_in, device=dev, dtype=dt)

        # fixed-step RK4 
        t = torch.linspace(0.0, 1.0, steps=64, device=dev, dtype=dt)
        vf = Velocity(self._vf_fn)
        trajectory = odeint(vf, x0, t, method="rk4")
        out = trajectory[-1]  # shape: [B, d_in] = [B, num + sum(K)]

        # prepare output: [num | cat]
        sample = torch.zeros(num_samples, d_out, device=dev, dtype=torch.float32)

        # 1) copy numericals
        num_end = self.num_numerical_features
        sample[:, :num_end] = out[:, :num_end].to(torch.float32)

        # 2) decode categorical from logits (the logits block starts at num_end)
        cat_logits_start = num_end
        idx = 0
        for i, K in enumerate(self.num_classes):
            logits_i = out[:, cat_logits_start + idx: cat_logits_start + idx + K]
            cls_i = torch.argmax(logits_i, dim=1)  
            sample[:, num_end + i] = cls_i.to(torch.float32) 
            idx += K

        return sample.cpu()`

but the resulting quality of the samples becomes very bad. I think I may be doing something wrong with the original code, could you give some insight? Thanks 

p.s. I am using a A100 so it shouldnt be an issue with compute power.

**Comment by babypink0524** [comment]
I've also encountered the issue of matrix multiplication dimension mismatch.

**Comment by kelvin715** [comment]
hey, I was wondering if you resolved this problem?

**Comment by andresguzco** [comment]
Hi @adiosbyebye, thank you so much for the detailed report and for your patience — I truly appreciate you taking the time to investigate this so thoroughly.

You were absolutely right about the `d_in` bug. This went unnoticed during the migration from JAX to PyTorch; the original JAX implementation handled the state dimensions differently, and the extra `+ len(self.num_classes)` term slipped in during the rewrite. Due to my workload, it took me longer than I would have liked to get around to addressing it.

This has now been fixed in commit `6214b13`, along with two related issues:

1. **`d_in` fix** (`flow_model.py:91`): Corrected to `d_in = self.num_numerical_features + sum(self.num_classes)`, exactly as you identified.

2. **Sampling index fix** (`flow_model.py:103-108`): The post-processing loop that maps ODE output to categorical columns had two bugs — `idx` started at `0` instead of `self.num_numerical_features`, and argmax results were written to columns `[0, 1, ...]` instead of `[d_numerical, d_numerical+1, ...]`. This was overwriting numerical columns with categorical values, which explains the poor sample quality you observed with your RK4 workaround.

3. **Velocity field fix**: The `Velocity` class now properly normalizes categorical logits through softmax before computing the velocity, which stabilizes the ODE integration and resolves the solver hanging issue you experienced with DOPRI5.

These fixes should resolve both the dimension mismatch error and the ODE solver stalling. The default DOPRI5 solver should now work as expected. Thank you again for your patience, and feel free to reopen if you run into anything else!

**Comment by andresguzco** [comment]
Resolved in commit 6214b13.
